File: frontend/src/lib/types/analysis.ts
export type Priority = "critical" | "important" | "normal" | "low";
export type Action = "review" | "reply" | "archive" | "delete_candidate" | "keep";
export type Bucket = "urgent" | "needs_reply" | "waiting" | "fyi";

export interface EmailAnalysisRequest {
  subject: string;
  sender: string;
  snippet?: string;
  body?: string | null;
}

export interface EmailAnalysisResult {
  summary: string;
  priority: string;
  category: string;
  action_items: string[];
}

export interface BulkEmailAnalysisRequest {
  emails: EmailAnalysisRequest[];
}

export interface AnalyzedEmail {
  original_email: EmailAnalysisRequest;
  analysis: EmailAnalysisResult;
}

export interface BulkEmailAnalysisResponse {
  results: AnalyzedEmail[];
}

export interface InboxSummaryResponse {
  total_emails: number;
  priority_counts: Record<string, number>;
  category_counts: Record<string, number>;
  top_action_items: string[];
  summary: string;
}

export interface EmailAnalysis {
  id: string;
  email_id: string;
  domain: string;
  subcategory: string;
  priority: Priority;
  action: Action;
  confidence: number;
  reason: string | null;
  model_name: string;
  schema_version: string;
  created_at: string;
}

File: backend/app/schemas/analysis.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class EmailAnalysisRequest(BaseModel):
    subject: str
    sender: str
    snippet: str = ""
    body: str | None = Field(
        default=None,
        description="Full extracted email body; preferred over snippet for analysis when present.",
    )


class EmailAnalysisResponse(BaseModel):
    summary: str
    priority: str
    category: str
    action_items: list[str]


BULK_EMAIL_ANALYSIS_MAX_SIZE = 50


class BulkEmailAnalysisRequest(BaseModel):
    emails: list[EmailAnalysisRequest] = Field(default_factory=list, max_length=BULK_EMAIL_ANALYSIS_MAX_SIZE)


class AnalyzedEmail(BaseModel):
    original_email: EmailAnalysisRequest
    analysis: EmailAnalysisResponse


class BulkEmailAnalysisResponse(BaseModel):
    results: list[AnalyzedEmail]


class InboxSummaryResponse(BaseModel):
    total_emails: int
    priority_counts: dict[str, int]
    category_counts: dict[str, int]
    top_action_items: list[str]
    summary: str


class EmailAnalysisCreate(BaseModel):
    domain: str
    subcategory: str
    priority: str
    action: str
    confidence: float = Field(ge=0.0, le=1.0)
    reason: str | None = None
    model_name: str = "mock-classifier-v1"
    schema_version: str = "v1"


class EmailAnalysisRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    email_id: str
    domain: str
    subcategory: str
    priority: str
    action: str
    confidence: float
    reason: str | None
    model_name: str
    schema_version: str
    created_at: datetime
